contracts/src/v0.8/dev/VRFCoordinatorV2.sol

```diff
@@ -0,0 +1,435 @@
+        assert(LINK.transfer(_recipient, _amount));
+    }
+
+    function getSubscription(uint64 subId)
+    external
+    view
+    returns (Subscription memory)
+    {
+        return s_subscriptions[subId];
+    }
+
+    function createSubscription(
+        address[] memory consumers // permitted consumers of the subscription
+    )
+    external
+    returns (uint64)
+    {
+        allConsumersValid(consumers);
+        currentSubId++;
+        s_subscriptions[currentSubId] = Subscription({
+            owner: msg.sender,
+            consumers: consumers,
+            balance: 0
+        });
+        emit SubscriptionCreated(currentSubId, msg.sender, consumers);
+        return currentSubId;
+    }
+
+    function allConsumersValid(address[] memory consumers)
+    internal
+    view
+    {
+        require(consumers.length <= s_config.maxConsumersPerSubscription, ">max consumers per sub");
+        for (uint i = 0; i < consumers.length; i++) {
+            require(consumers[i] != address(0), "consumer add == zero");
+        }
+    }
+
+    function updateSubscription(
+        uint64 subId,
+        address[] memory consumers // permitted consumers of the subscription
+    )
+    external
+    {
+        require(msg.sender == s_subscriptions[subId].owner, "sub owner must update");
+        allConsumersValid(consumers);
+        address[] memory oldConsumers = s_subscriptions[subId].consumers;
+        s_subscriptions[subId].consumers = consumers;
+        emit SubscriptionConsumersUpdated(subId, oldConsumers, consumers);
+    }
+
+    function fundSubscription(
+        uint64 subId,
+        uint256 amount
+    )
+    external
+    {
+        require(s_subscriptions[subId].owner != address(0), "subID doesnt exist");
+        require(msg.sender == s_subscriptions[subId].owner, "sub owner must fund");
+        uint256 oldBalance = s_subscriptions[subId].balance;
+        s_subscriptions[subId].balance += amount;
+        LINK.transferFrom(msg.sender, address(this), amount);
+        emit SubscriptionFundsAdded(subId, oldBalance, s_subscriptions[subId].balance);
+    }
+
+    function withdrawFromSubscription(
+        uint64 subId,
+        address to,
+        uint256 amount
+    )
+    external
+    {
+        require(msg.sender == s_subscriptions[subId].owner, "sub owner must withdraw");
+        require(s_subscriptions[subId].balance >= amount, "insufficient balance");
+        uint256 oldBalance = s_subscriptions[subId].balance;
+        s_subscriptions[subId].balance -= amount;
+        LINK.transfer(to, amount);
+        emit SubscriptionFundsWithdrawn(subId, oldBalance, s_subscriptions[subId].balance);
+    }
+
+    // Keep this separate from zeroing, perhaps there is a use case where consumers
+    // want to keep the subId, but withdraw all the link.
+    function cancelSubscription(
+        uint64 subId
+    )
+    external
+    {
+        require(msg.sender == s_subscriptions[subId].owner, "sub owner must cancel");
```

Comment:
I think the [new custom error types](https://blog.soliditylang.org/2021/04/21/custom-errors/) are supposed to reduce contract space, and are now supported by Hardhat. Probably worth switching to them.